### PR TITLE
feat(kudoctl): request client

### PR DESCRIPTION
**Pull request by NilsPonsard**
This PR adds a wrapper to ease the requests to the kudo controller.

### CODE REVIEW

File: kudoctl/src/request.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        Ok(Client {
+            client: reqwest::Client::builder()
+                .default_headers(headers)
+                .build()?,
+            base_url,
+        })
+    }
+
+    // Do a request to the controller.
+    pub async fn do_request(
```
Comment (sameo) [inline kudoctl/src/request.rs:37]: `send_request()` would be better.

File: kudoctl/src/request.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    ) -> Result<Response, Box<dyn Error>> {
+        let url = self.base_url.join(endpoint)?;
+        let mut request = (*self).client.request(method, url);
+
+        if let Some(body) = body {
+            request = request.body(body.to_owned());
+        }
+
+        let result = request.send().await?;
+        Ok(result)
```
Comment (sameo) [inline kudoctl/src/request.rs:51]: Just `request.send().await`

File: kudoctl/src/request.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    // Do a request to the controller and deserialize the response.
+    pub async fn do_json_request<T: DeserializeOwned>(
```
Comment (sameo) [inline kudoctl/src/request.rs:55]: `send_json_request`

File: kudoctl/src/request.rs (hunk 1 above)
Comment (sameo) [inline kudoctl/src/request.rs:37]: I would keep that one private for now.

File: kudoctl/src/request.rs (hunk 2 above)
Comment (NilsPonsard) [reply kudoctl/src/request.rs:51]: When i tried there was a problem with the return type of the function being a Box<dyn Error> and `request.send()` returning another type of error and I can’t figure out the solution.

Comment (sameo) [reply kudoctl/src/request.rs:51]: Try:

```rust
request.send().await.map_err(|e| Box::new(e) as Box<dyn Error>)
```

**Comment by NilsPonsard** [comment]
Okay this should work now

**Comment by sameo** [comment]
> Okay this should work now

Thanks. Please apply the same logic to `send_json_request` and we'll get that PR merged.

**Comment by NilsPonsard** [comment]
Didn’t think of that, thanks. This should be good now.

**Review by sameo — APPROVED** [review]

**Pull request merged**